src/vs/editor/common/languages.ts

```diff
@@ -176,11 +192,33 @@ export interface Hover {
  */
 export interface HoverProvider {
 	/**
-	 * Provide a hover for the given position and document. Multiple hovers at the same
+	 * Provide a hover for the given position, context and document. Multiple hovers at the same
 	 * position will be merged by the editor. A hover can have a range which defaults
 	 * to the word range at the position when omitted.
 	 */
-	provideHover(model: model.ITextModel, position: Position, token: CancellationToken): ProviderResult<Hover>;
+	provideHover(model: model.ITextModel, position: Position, token: CancellationToken, context?: HoverContext): ProviderResult<Hover>;
+}
+
+export interface HoverContext {
+	/**
+	 * Whether to increase or decrease the hover's verbosity
+	 */
+	action: HoverVerbosityAction;
+	/**
+	 * The previous hover for the same position
+	 */
+	previousId: string;
```

Comment:
I wouldn't use an `id` here.
I would make `HoverProvider` generic (`HoverProvider<THover extends Hover = Hover>`) and pass in `previous: THover`

Reply:
So I was following the pattern which is used for releasing code actions items (used elsewhere too) here https://github.com/microsoft/vscode/blob/53f1f141fb637c33883c5914b982214b36eaf8cf/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L549. At this position https://github.com/microsoft/vscode/blob/53f1f141fb637c33883c5914b982214b36eaf8cf/src/vs/workbench/api/browser/mainThreadLanguageFeatures.ts#L381, the code action cache ID obtained from `$provideCodeActions` is passed in to `$releaseCodeActions`

Reply:
We are utlimately only interested in the hover ID to retrieve the previous hover in the adapter, would it not be redundant to send the whole object?

Reply:
Please only *Resolve* a conversation when it is done and all questions are qualified. As a reviewer, I usually don't go through resolved conversations again, but assume they are taken care of.

Ultimately, an id has to be used. But editor-core APIs should not use string ids to reference objects, but just object references directly. If you call a method, it does not make a difference if you pass in `obj.id` vs `obj`.
`obj.id` is just confusing and might cause bugs.

To get some inspiration, here is how I did it with inline completions:

https://github.com/microsoft/vscode/blob/3e9c26d799819ba37845cbfb1d44393fea0c6691/src/vs/editor/common/languages.ts#L733

And then I implemented it like this:

https://github.com/microsoft/vscode/blob/3e9c26d799819ba37845cbfb1d44393fea0c6691/src/vs/workbench/api/browser/mainThreadLanguageFeatures.ts#L601

`IdentifiableInlineCompletions` has `pid` to make it possible to send over the list by reference.

Reply:
The reason I resolved the conversation is because after thinking some more, I made the change. I pass in the original hover, and then indeed pass in the id of the object into $providerHover

Yes the explanation makes sense